pkg/plugins/policies/meshaccesslog/plugin/v1alpha1/plugin.go

```diff
@@ -308,52 +300,114 @@ func configureListener(
+		return BaseAccessLogBuilder(b, defaultFormat, backendsAcc, kumaValues, accessLogSocketPath).
+			ConfigureIf(r.Protocol.IsHTTPBased(), func() Configurer[envoy_accesslog.AccessLog] {
+				return bldrs_al.MetadataFilter(true, bldrs_matcher.NewMetadataBuilder().
+					Configure(bldrs_matcher.Key(envoy_wellknown.FileAccessLog, routeMetadataKey)).
+					Configure(bldrs_matcher.NullValue()))
+			})
+	}
+
+	builderForRouteBackend := func(routeID kri.Identifier) func(b api.Backend) *Builder[envoy_accesslog.AccessLog] {
+		return func(b api.Backend) *Builder[envoy_accesslog.AccessLog] {
+			return BaseAccessLogBuilder(b, defaultFormat, backendsAcc, kumaValues, accessLogSocketPath).
+				Configure(bldrs_al.MetadataFilter(false, bldrs_matcher.NewMetadataBuilder().
+					Configure(bldrs_matcher.Key(envoy_wellknown.FileAccessLog, routeMetadataKey)).
+					Configure(bldrs_matcher.ExactValue(routeID.String()))))
 		}
+	}
 
-		for _, chain := range listener.FilterChains {
-			if err := listeners_v3.UpdateHTTPConnectionManager(chain, func(hcm *envoy_hcm.HttpConnectionManager) error {
-				hcm.AccessLog = append(hcm.AccessLog, accessLog)
-				return nil
-			}); err != nil {
-				return err
-			}
-			if err := listeners_v3.UpdateTCPProxy(chain, func(tcpProxy *envoy_tcp.TcpProxy) error {
-				tcpProxy.AccessLog = append(tcpProxy.AccessLog, accessLog)
-				return nil
-			}); err != nil {
-				return err
-			}
+	return NewModifier(listener).
+		Configure(bldrs_listener.AccessLogs(util_slices.Map(backends, builderForSharedBackend))).
+		Configure(bldrs_listener.Routes(
+			util_maps.MapValues(
+				routesConfs,
+				func(id kri.Identifier, _ api.Conf) Configurer[routev3.Route] {
+					return bldrs_route.Metadata(routeMetadataKey, id.String())
+				}))).
+		Configure(bldrs_listener.AccessLogs(
+			util_maps.FlatMapKV(
+				routesConfs,
+				func(id kri.Identifier, conf api.Conf) []*Builder[envoy_accesslog.AccessLog] {
+					return util_slices.Map(
+						pointer.Deref(conf.Backends),
+						builderForRouteBackend(id),
+					)
+				}))).
+		ConfigureIf(hasAtLeastOneBackend, func() Configurer[envoy_listener.Listener] {
+			return bldrs_listener.HCM(hcm.LuaFilterAddFirst(luaFilter))
+		}).
+		Modify()
+}
+
+func buildRoutesMap(l *envoy_listener.Listener, svcCtx *outbound.ResourceContext[api.Conf]) (map[kri.Identifier]api.Conf, error) {
+	routes := map[kri.Identifier]api.Conf{}
+	if err := bldrs_listener.TraverseRoutes(l, func(route *routev3.Route) {
+		if !kri.IsValid(route.Name) {
+			return
+		}
+		id, _ := kri.FromString(route.Name)
+		if conf, isDirect := svcCtx.WithID(id).DirectConf(); isDirect {
+			routes[id] = conf
 		}
+	}); err != nil {
+		return nil, err
 	}
-	return nil
+	return routes, nil
 }
+
+const routeMetadataKey = "route_kri"
+
+var luaFilter = `function envoy_on_request(handle)
+  local meta = handle:metadata():get("route_kri")
+  if meta ~= nil then
+    handle:streamInfo():dynamicMetadata():set("envoy.access_loggers.file", "route_kri", meta)
```

Comment:
Can we template it with `routeMetadataKey` const to avoid potential issue with changing the name?

Reply:
added proper template with params